### PR TITLE
backupengine: send periodic progress updates

### PR DESCRIPTION
**Pull request by vmg**
## Description

Fixes https://github.com/vitessio/vitess/issues/9195

As explained on the linked issue, whenever a very large database is being backed up, the GRPC connection can time out. This is despite the fact that the GRPC request we're performing has been designed to stream back the progress of a backup. The reason why the timeout occurs is because the copying of individual files in a backup is performed via `io.Copy`, which is essentially a busy-loop for copying data, so there is no actual progress report being sent via GRPC.

To work around this, I've implemented a custom `io.Writer` called `backupWriter`, which takes care of both hashing the backup as it is being written, and of keeping tack of how much of the file has been written, so it is capable of sending progress reports via GRPC while the copy phase is happening.

This applies both when performing a backup and when restoring one, so we're killing two birds with one stone in this PR.

## Related Issue(s)
- https://github.com/vitessio/vitess/issues/9195

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 1]
```diff
@@ -602,13 +664,13 @@ func (be *BuiltinBackupEngine) restoreFile(ctx context.Context, params RestorePa
...
 	// Check the hash.
-	hash := hasher.HashString()
+	hash := bw.HashString()
 	if hash != fe.Hash {
 		return vterrors.Errorf(vtrpc.Code_INTERNAL, "hash mismatch for %v, got %v expected %v", fe.Name, hash, fe.Hash)
 	}
 
 	// Flush the buffer.
-	if err := dst.Flush(); err != nil {
+	if err := bw.Close(); err != nil {
```
Comment (deepthi) [inline go/vt/mysqlctl/builtinbackupengine.go:673]: nit: fix error message below to match the code change.

**Review by deepthi — APPROVED** [review]
Looks fine to me, but some of the backup/restore tests are failing.

**Comment by vmg** [comment]
@deepthi Green now :ok_hand:

File: go/vt/mysqlctl/builtinbackupengine.go (hunk 1 above)
Comment (deepthi) [reply go/vt/mysqlctl/builtinbackupengine.go]: Sorry to be nitpicky, but I don't think this was addressed? It's line 540 now.

Comment (vmg) [reply go/vt/mysqlctl/builtinbackupengine.go]: It had been addressed when you wrote the comment, I think you may have been looking at the wrong line. And you didn't approve the PR. :sweat:

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 2]
```diff
@@ -446,13 +535,13 @@ func (be *BuiltinBackupEngine) backupFile(ctx context.Context, params BackupPara
 		}
 	}
 
-	// Flush the buffer to finish writing on destination.
-	if err = dst.Flush(); err != nil {
+	// Close the backupPipe to finish writing on destination.
+	if err = bw.Close(); err != nil {
 		return vterrors.Wrapf(err, "cannot flush destination: %v", name)
 	}
```
Comment (deepthi) [inline go/vt/mysqlctl/builtinbackupengine.go:541]: This is what I was referring to.
The comment says `Close the backupPipe` and the error message says `cannot flush destination`.
I _want_ to approve and merge this PR!
I'll approve provisionally and you can fix the error message and merge it.

**Review by deepthi — APPROVED** [review]
Nice fix!

**Pull request merged**